library/src/scala/main.scala

```diff
@@ -8,6 +8,192 @@
 
 package scala
 
+import collection.mutable
+
+
 /** An annotation that designates a main function
  */
-class main extends scala.annotation.Annotation {}
+class main extends scala.annotation.MainAnnotation:
+  self =>
+  import main._
+
+  protected sealed abstract trait Argument {
+    def name: String
+    def typeName: String
+    def doc: String
+
+    def usage: String
+  }
+  protected class SimpleArgument(val name: String, val typeName: String, val doc: String) extends Argument:
+    override def usage: String = s"[--$name] <$name>"
+  protected class OptionalArgument[T](val name: String, val typeName: String, val doc: String, val defaultValue: T)
+    extends Argument:
+    override def usage: String = s"[[--$name] <$name>]"
+  protected class VarArgument(val name: String, val typeName: String, val doc: String) extends Argument:
+    override def usage: String = s"[<$name> [<$name> [...]]]"
+
+  override type ArgumentParser[T] = util.CommandLineParser.FromString[T]
+  override type MainResultType = Any
+
+  /** Prints the main function's usage */
+  def usage(commandName: String, args: Seq[Argument]): Unit =
+    def wrappedArgumentUsages(argsUsage: List[String], maxLength: Int): List[String] =
+      def recurse(args: List[String], currentLine: String, acc: Vector[String]): Vector[String] =
+        (args, currentLine) match {
+          case (Nil, "") => acc
+          case (Nil, l) => (acc :+ l)
+          case (arg :: t, "") => recurse(t, arg, acc)
+          case (arg :: t, l) if l.length + 1 + arg.length <= maxLength => recurse(t, s"$l $arg", acc)
+          case (arg :: t, l) => recurse(t, arg, acc :+ l)
+        }
+
+      recurse(argsUsage, "", Vector()).toList
+    end wrappedArgumentUsages
+
+    val maxLineLength = 120 // TODO as parameter? As global Dotty parameter?
+
+    val usageBeginning = s"Usage: $commandName "
+    val argsOffset = usageBeginning.length
+    val argUsages = wrappedArgumentUsages(args.map(_.usage).toList, maxLineLength - argsOffset)
+
+    println(usageBeginning + argUsages.mkString("\n" + " " * argsOffset))
+
+  /** Prints an explanation about the function */
+  def explain(commandName: String, commandDoc: String, args: Seq[Argument]): Unit =
+    if (commandDoc.nonEmpty)
+      println(commandDoc)
+    if (args.nonEmpty) {
+      val argNameShift = 2
+      val argDocShift = argNameShift + 2
+
+      println("Arguments:")
+      for (arg <- args)
+        val argDoc = StringBuilder(" " * argNameShift)
+        argDoc.append(s"${arg.name} - ${arg.typeName}")
+
+        arg match {
+          case o: OptionalArgument[?] => argDoc.append(" (optional)")
+          case _ =>
+        }
+
+        if (arg.doc.nonEmpty) {
+          val shiftedDoc = arg.doc.split("\n").map(" " * argDocShift + _).mkString("\n")
```

Comment:
I am still having the typing issue we spoke about on monday. It did not appear in my compiler before I merged the recent changes on lampepfl/dotty/master.
The error is:
```
[error] -- [E008] Not Found Error: D:\Bureau\dotty\library\src\scala\main.scala:80:47 --
[error] 80 |          val shiftedDoc = arg.doc.split("\n").map(" " * argDocShift + _).mkString("\n")
[error]    |                           ^^^^^^^^^^^^^^^^^^^^^^^
[error]    |      value map is not a member of Array[String | Null] | Null.
[error]    |      An extension method was tried, but could not be fully constructed:
[error]    |
[error]    |          refArrayOps[T](arg.doc.split("\n"))
```

Reply:
This was caused by the change in #13729.

It can be fixed with
```diff
-          val shiftedDoc = arg.doc.split("\n").map(" " * argDocShift + _).mkString("\n")
+          val shiftedDoc = arg.doc.split("\n").nn.map(" " * argDocShift + _).mkString("\n")
```